packages/react-reconciler/src/ReactFiberWorkLoop.new.js

```diff
@@ -1539,19 +1366,23 @@ export function renderHasNotSuspendedYet(): boolean {
   return workInProgressRootExitStatus === RootIncomplete;
 }
 
-function renderRootSync(root, expirationTime) {
+function renderRootSync(root: FiberRoot, lanes: Lanes) {
   const prevExecutionContext = executionContext;
   executionContext |= RenderContext;
   const prevDispatcher = pushDispatcher(root);
 
-  // If the root or expiration time have changed, throw out the existing stack
+  // If the root or lanes have changed, throw out the existing stack
   // and prepare a fresh one. Otherwise we'll continue where we left off.
-  if (
-    root !== workInProgressRoot ||
-    !isSameExpirationTime(expirationTime, renderExpirationTime)
-  ) {
-    prepareFreshStack(root, expirationTime);
-    startWorkOnPendingInteractions(root, expirationTime);
+  if (workInProgressRoot === null) {
+    prepareFreshStack(root, lanes);
+    startWorkOnPendingInteractions(root, lanes);
+  } else if (workInProgressRoot !== root) {
+    workInProgressRoot.wipLanes = NoLanes;
+    prepareFreshStack(root, lanes);
+    startWorkOnPendingInteractions(root, lanes);
+  } else if (lanes !== root.wipLanes) {
+    startWorkOnPendingInteractions(root, lanes);
+    prepareFreshStack(root, lanes);
```

Comment:
Calling `prepareFreshStack` *after* doing something looks odd. Can you explain why we do it?

Reply:
We call both of these function right before starting a new render at a new level. I don't think the order matters since `prepareFreshStack` doesn't touch `root.memoizedInteractions`, but I'll rearrange them for consistency with the other branches.